lib/web/users_test.go

```diff
@@ -118,6 +119,186 @@ func TestCRUDs(t *testing.T) {
+		{
+			name: "Deduplicates",
+			updateReq: saveUserRequest{
+				Name:   "deduplicates",
+				Roles:  defaultRoles,
+				Logins: &[]string{"login1", "login2", "login1"},
+			},
+			expectedTraits: map[string][]string{
+				teleport.TraitLogins: {"login1", "login2"},
+			},
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			t.Parallel()
+			user, err := types.NewUser(tt.name)
+			require.NoError(t, err)
+			user.SetRoles(defaultRoles)
+
+			m := &mockedUserAPIGetter{}
+			m.mockGetUser = func(name string, withSecrets bool) (types.User, error) {
+				return user, nil
+			}
+			m.mockUpdateUser = func(ctx context.Context, user types.User) error {
+				user.SetTraits(user.GetTraits())
+				return nil
+			}
+
+			_, err = updateUser(newRequest(t, tt.updateReq), m, "")
+			require.NoError(t, err)
+
+			for traitName, traitList := range tt.expectedTraits {
+				require.ElementsMatch(t, traitList, user.GetTraits()[traitName])
+			}
+
+			require.ElementsMatch(t, user.GetRoles(), defaultRoles)
+		})
+	}
+}
+
+func TestSetTraitDoesntRemoveOtherTraits(t *testing.T) {
+	username := "user1"
+	defaultRoles := []string{"role1"}
+
+	updateReq := saveUserRequest{
+		Name:      username,
+		Roles:     defaultRoles,
+		KubeUsers: &[]string{"kubeuser1", "kubeuser2"},
+	}
+
+	user, err := types.NewUser(username)
+	require.NoError(t, err)
+	user.SetRoles(defaultRoles)
+	user.SetTraits(map[string][]string{
+		teleport.TraitLogins:  {"login1"},
+		teleport.TraitDBNames: {"dbname1", "dbname2"},
+	})
+
+	expectedTraits := map[string][]string{
+		teleport.TraitKubeUsers: {"kubeuser1", "kubeuser2"},
+		teleport.TraitLogins:    {"login1"},
+		teleport.TraitDBNames:   {"dbname1", "dbname2"},
+	}
+
+	m := &mockedUserAPIGetter{}
+	m.mockGetUser = func(name string, withSecrets bool) (types.User, error) {
+		return user, nil
+	}
+	m.mockUpdateUser = func(ctx context.Context, user types.User) error {
+		user.SetTraits(user.GetTraits())
+		return nil
+	}
+
+	_, err = updateUser(newRequest(t, updateReq), m, "")
+	require.NoError(t, err)
+
+	for traitName, traitList := range expectedTraits {
+		require.ElementsMatch(t, traitList, user.GetTraits()[traitName])
+	}
+}
+
+func TestSetTraitCanRemoveTraits(t *testing.T) {
+	username := "user1"
+	defaultRoles := []string{"role1"}
+
+	updateReq := saveUserRequest{
+		Name:   username,
+		Roles:  defaultRoles,
+		Logins: &[]string{},
+	}
+
+	user, err := types.NewUser(username)
+	require.NoError(t, err)
+	user.SetRoles(defaultRoles)
+	user.SetTraits(map[string][]string{
+		teleport.TraitLogins: {"login1"},
+	})
+
+	m := &mockedUserAPIGetter{}
+	m.mockGetUser = func(name string, withSecrets bool) (types.User, error) {
+		return user, nil
+	}
+	m.mockUpdateUser = func(ctx context.Context, user types.User) error {
+		user.SetTraits(user.GetTraits())
+		return nil
+	}
+
+	_, err = updateUser(newRequest(t, updateReq), m, "")
+	require.NoError(t, err)
+
+	require.Empty(t, user.GetTraits()[teleport.TraitLogins])
```

Comment:
I think we could combine all 3 tests in a single table, or at least combine this and the previous one, as the setup is rather similar. Up to you.

Reply:
Done :+1: